pydantic/_internal/_generate_schema.py

```diff
@@ -766,3 +771,137 @@ def get_model_self_schema(cls: type[BaseModel]) -> core_schema.ModelSchema:
         core_schema.definition_reference_schema(model_ref),
         metadata=build_metadata_dict(js_metadata=model_js_metadata),
     )
+
+
+def apply_discriminator(schema: core_schema.CoreSchema, discriminator: str) -> core_schema.CoreSchema:
+    # Eventually: should add support for other discriminator types, and explicitly specified choices
+
+    # If the field was wrapped with a nullable schema, we preserve that here and re-wrap the final result at the end.
+    nullable_schema: core_schema.NullableSchema | None = None
+    if schema['type'] == 'nullable':
+        nullable_schema = schema
+        schema = schema['schema']
+
+    if schema['type'] != 'union':
+        raise TypeError('`discriminator` can only be used with `Union` type with more than one variant')
+
+    union_choices = schema['choices']
+    if len(union_choices) < 2:
+        raise TypeError('`discriminator` can only be used with `Union` type with more than one variant')
+
+    # `aliases` is meant to behave like a set, but uses a dict to ensure order is preserved
+    aliases = {discriminator: None}
+    tagged_union_choices = _build_tagged_union_choices(union_choices, discriminator, aliases)
+
+    if len(aliases) > 1:
+        schema_discriminator: str | list[list[str | int]] = [[alias] for alias in aliases]
+    else:
+        schema_discriminator = discriminator
+
+    discriminated_schema = core_schema.tagged_union_schema(
+        choices=tagged_union_choices,
+        discriminator=schema_discriminator,
+        custom_error_type=schema.get('custom_error_type'),
+        custom_error_message=schema.get('custom_error_message'),
+        custom_error_context=schema.get('custom_error_context'),
+        strict=False,
+        ref=schema.get('ref'),
+        metadata=schema.get('metadata'),
+        serialization=schema.get('serialization'),
+    )
+
+    if nullable_schema is None:
+        return discriminated_schema
+    else:
+        return core_schema.nullable_schema(
+            discriminated_schema,
+            strict=nullable_schema.get('strict'),
+            ref=nullable_schema.get('ref'),
+            metadata=nullable_schema.get('metadata'),
+            serialization=nullable_schema.get('serialization'),
+        )
+
+
+def _build_tagged_union_choices(
+    choices: typing.Iterable[core_schema.CoreSchema], discriminator: str, aliases: dict[str, None]
+) -> dict[str | int, str | int | core_schema.CoreSchema]:
+    choices = list(choices)[::-1]
```

Comment:
I am purposely making a copy of the list so I can pop items out below. I reversed it so that the "popping" below produces items in the same order they occurred in the list (since they will pop off the end instead of from the start).

That said, writing this, I realize I need to reverse the items below while extending in the coalescing case.

Reply:
I would still accept a list an make a copy. It may not be a problem here but IMO the issue with this sort of API is that the caller may assume you're being lazy processing the items and thus give you `range(1, 1_000_000_000_000)` but then if you try to put that all into memory kaboom.

Reply:
Also reversing the list will copy it:

```
>>> ls = [1, 2, 3]
>>> rv = ls[::-1]
>>> rv.pop()
1
>>> ls
[1, 2, 3]
```

Reply:
I realized that too, but sadly also discovered this mis-typing issue: https://github.com/pydantic/pydantic-core/issues/427